File: solutions/elixir/wordy/1/lib/wordy.ex
defmodule Wordy do
  @doc """
  Calculate the math problem in the sentence.
  """
  @spec answer(String.t()) :: integer
  def answer(question) do
    parse(String.split(question))
  end

  def parse(["What", "is", number1, operator, number2]) do
    {num1, num2} = parse_numbers(number1, number2)
    case operator do
      "plus" -> num1 + num2
      "minus" -> num1 - num2
    end
  end

  def parse(["What", "is", number1, operator1, number2, operator2, number3]) do
    {num1, num2} = parse_numbers(number1, number2)
    temp = case operator1 do
      "plus" -> num1 + num2
      "minus" -> num1 - num2
    end
    num3 = parse_number(number3)
    case operator2 do
      "plus" -> temp + num3
      "minus" -> temp - num3
    end
  end

  def parse(["What", "is", number1, operator1, number2, operator2, "by", number3]) do
    {num1, num2} = parse_numbers(number1, number2)
    temp = case operator1 do
      "plus" -> num1 + num2
      "minus" -> num1 - num2
    end
    num3 = parse_number(number3)
    case operator2 do
      "multiplied" -> temp * num3
      "divided" -> div(temp, num3)
    end
  end

  def parse(["What", "is", number1, operator, "by", number2]) do
    {num1, num2} = parse_numbers(number1, number2)
    case operator do
      "multiplied" -> num1 * num2
      "divided" -> div(num1, num2)
    end
  end

  def parse(["What", "is", number1, operator1, "by", number2, operator2, "by", number3]) do
    {num1, num2} = parse_numbers(number1, number2)
    temp = case operator1 do
      "multiplied" -> num1 * num2
      "divided" -> div(num1, num2)
    end
    num3 = parse_number(number3)
    case operator2 do
      "multiplied" -> temp * num3
      "divided" -> div(temp, num3)
    end
  end

  def parse(["What", "is", number]) do
    parse_number(number)
  end

  def parse(_), do: raise(ArgumentError)

  def parse_number(number) do
    number
    |> String.replace("?", "")
    |> String.to_integer()
  end

  def parse_numbers(number1, number2) do
    num1 = parse_number(number1)
    num2 = parse_number(number2)
    {num1, num2}
  end
end

File: solutions/csharp/wordy/1/Wordy.cs
using System.IO.Pipelines;
using System.Text.RegularExpressions;

public static class Wordy
{
    public static int Answer(string question)
    {

        try
        {
            var threeOperandTest = Regex.Match(question, @"What is (?<Op1>-?\d+) (?<Operator1>\b(divided by|plus|multiplied by|minus)\b) (?<Op2>-?\d+) (?<Operator2>\b(divided by|plus|multiplied by|minus)\b) (?<Op3>-?\d+)\?");

            if (threeOperandTest.Success)
            {
                var operand1 = Int32.Parse(threeOperandTest.Groups["Op1"].Value);
                var operand2 = Int32.Parse(threeOperandTest.Groups["Op2"].Value);
                var operand3 = Int32.Parse(threeOperandTest.Groups["Op3"].Value);
                var op1 = threeOperandTest.Groups["Operator1"].Value;
                var op2 = threeOperandTest.Groups["Operator2"].Value;

                var result = ResolveTwoOperandEquation(operand1, operand2, op1);

                return ResolveTwoOperandEquation(result, operand3, op2);
            }

            var twoOperandTest = Regex.Match(question, @"What is (?<Op1>-?\d+) (?<Operator>\b(divided by|plus|multiplied by|minus)\b) (?<Op2>-?\d+)\?");

            if (twoOperandTest.Success)
            {
                var operand1 = Int32.Parse(twoOperandTest.Groups["Op1"].Value);
                var operand2 = Int32.Parse(twoOperandTest.Groups["Op2"].Value);
                var op = twoOperandTest.Groups["Operator"].Value;
                return  ResolveTwoOperandEquation(operand1, operand2, op);
            }

            var identityTest = Regex.Match(question, @"What is (?<Value>\d+)\?");

            if (identityTest.Success)
            {
                return Int32.Parse(identityTest.Groups["Value"].Value);
            }

        }
        catch (System.FormatException)
        {
            throw new ArgumentException();
        }   
             
        throw new ArgumentException();
    }

    private static int ResolveTwoOperandEquation(int operand1, int operand2, string op)
    {
        if (op == "divided by")
        {
            return operand1 / operand2;
        }
        if (op == "plus")
        {
            return operand1 + operand2;
        }
        if (op == "multiplied by")
        {
            return operand1 * operand2;
        }
        if (op == "minus")
        {
            return operand1 - operand2;
        }

        throw new ArgumentException();
    }
}